Several Improvements to the status bar. (multi-sections, text markup, unicode icons)

cmanv opened the pull request:
This is a proposed enhancement for the status bar. It is enabled with bar_split = 1

This will split the bar text into 3 parts:
- The left part which is left justified.
- The right part which is right justified.
- The middle part which occupies the free space between the two and is centered within that space.

The split is done with the key sequences +\ and +/ within the bar_format.
+\  marks the start of the middle text
+/ marks the start of the right text 

Example of a split bar format:
bar_format = +L +\ +W +/ +A

In addition, 4 color resources are defined  to control the background and foreground color of the middle add right part:
- bar_color_middle
- bar_color_right
- bar_font_color_middle
- bar_font_color_right

The left part use the default resources. 
Also the value of bar_justify is ignored when bar_split is set.

cmanv commented:
Close this for now as this patch has now grown into something more ambitious.

LordReg commented:
Been using this for a while now, apart from a few minor tweaks I find this solution acceptable since it is simple, backwards compatible, etc.  Nice patch.

What improvements/enhancements to this patch do you have in mind?
A few improvements that cross my mind:
- A single delimiter such as `+|` that allows for n-sections with ability to specify alignment for each section.
- A toggle shouldn't be needed so long as the delimiter begins with `+`.
- Colors should be handled similar to color_focus/color_focus_maximized, etc.  When bar_color is set, bar colors not specified by the user should follow.
- Fix EOL white-space.

cmanv commented:
Nice to see someone tried it. I'll post the new patch in a couple of days.
This new version will include:
- The ability to markup the bar text. (font, text color and background color)
- Support for Material Design Icons, including for the workspace names.
- Switch to a single delimiter +|  for splitting the bar. (you read my mind ?)
- Of course I intend to keep these changes 100% backward compatible.

As for your comment:
- I've used a toggle because currently the split mode disables the bar_justify setting. It is not backward compatible.
- Right now, justification could make sense for the center section, as it is the one section that gets all the remaining free space. The left and right sections just use what they need. Some mechanism would be needed to allocate space to each section thus allowing to enable justification per section. I'll think about that.

LordReg commented:
It isn't explicitly specified in the man page, but undefined `+` character sequences for `bar_format` are reserved for future use.  As stated, undefined sequences are copied as-is but the user should always use `++` to specify a literal `+` instead of relying on such behavior.  I don't see having a new character sequence changing some default behavior as breaking backwards compatibility.

BTW, having `bar_justify` apply to the center section was actually one of the "minor tweaks" that came to mind.

Looking forward to checking out the new patch.

cmanv commented:
This pull request now contains several improvement to the status bar.

1- The bar layout is now handled by the **bar_layout** setting.

The format consist of a list of letters (C,L,R,T) separated by "|", for example: 
**T|L|C|R**
A 'T' section has a width determined by the length of the text. The 'L', 'C' and 'R' sections are respectively left, center and right-justified. The 'L', 'C' and 'R' sections equally share the horizontal space not used by the 'T' sections. This may be changed by prefixing the letter with an integer. For example:  
**2L|3C|5L**
The available space is then allocated according to the prefix  which act as a weight of the
sections. The default weight of a non-T section is 1. If **bar_layout** is not defined or malformed, the default is to create only one section justified according to the value of **bar_justify** .

2- The text font, the text color and the background color can be change with markup sequences. (Xft font only)

These 3 existing resources have been modified to accept up to 10 comma separated values:
**bar_font**
**bar_color**
**bar_font_color**

In each case, the first value on the list is the default, and the behavior is the same as before.But now, the text font, text color and background color can now be changed in the status bar with these markup sequences:

+@fn0, +@fn1, .., +@fn9  select which font is used.
+@fg0, +@fg1, .., +@bg9  select the text color
+@bg0, +@bg1, .., +@bg9  select the background color.
+@stp is used to stop the parsing of markup sequences. From that point, any markup sequence will be treated as plain text. This can be useful to avoid unexpected highlighting in free format text, or for debugging  text from baraction.sh.

 Note that the effect of markup sequences do not cross section boundaries.

3- Support for unicode font icons. (Xft only)

A new resources has been created to specify a unicode font for icons.
**bar_font_icons**
This allow the display of icons from 3-bytes unicode characters. 
(Successfully tested with the 'Material Design Icons' font)

cmanv commented:
New update to this patch, mainly concerning legacy fonts.

- Allow the use of color markup sequence with legacy fonts. There was no reason not to.
- Update the manpage to reflect this change.
- Replace all instances of XmbTextExtents with a new macro TEXTEXTENTS(), working 
  alongside DRAWSTRING(). This allows a consistent use of  Xmb* or Xutf8* functions. 
- Minor correction in bar_setup_layout

cmanv commented:
Changes from this last commit:
- The sequence +A is now replaced before the bar format is divided into sections. As a consequence, the external script can now have multiple sections in its output. In fact, all '+' sequences from the bar format can be in the scripts output (except +A).
- Removed and replaced all regexec() from the markup sequence parsing code. This noticeably reduces the CPU usage of the process.

marcopeereboom commented:
Does this leave the defaults alone? if so I am ok with it going in.

cmanv commented:
Yes defaults have been kept as is. There are no user visible changes if the config file is not modified to use the new features.

marcopeereboom commented:
@LordReg?

vext01 commented:
Sounds like a cool idea. I've always wanted the time on the right-hand side of the bar.

marcopeereboom commented:
I need to know that this is 100% backwards compat and doesn't change appearance for current users. If that is the case I am ok with it.

LordReg commented:
@cmanv I'm thinking bar_layout could be integrated into bar_format to eliminate the need to synchronize the two options.
New syntax: `+|<weight><mode>`
Where 'weight' is optional and 'mode' required as in your `bar_layout` syntax
Marking of the first section would be optional.  If it is omitted, use default weight (1) and align according to `bar_justify` for that section.

Example:
```
# Old syntax
bar_format = +S[+L]{+M}+F+|+A+W+|%a %y-%m-%d %R
bar_layout = T|C|T

# New syntax
bar_format = +|T+S[+L]{+M}+F+|C+A+W+|T%a %y-%m-%d %R
```
Also,
spectrwm aims to follow [OpenBSD style(9)](https://man.openbsd.org/style) and 80col (with a few exceptions where it would make things hard to read.)
mandoc syntax: Sentences should terminate at the end of an input line.
https://man.openbsd.org/roff.7#Sentence_Spacing

cmanv commented:
@LordReg thanks for your comment. I'll look into that.

cmanv commented:
As suggested by LordReg, drop the **bar_layout** option and integrate the layout specification into **bar_format**. The start of a section in  **bar_format** is now marked by:

+|[weight][justify] 

where **weight** is a positive integer (default 1) and **justify** has one of the values T,L,C,R. (default is taken from the value of the **bar_justify** option)

The man page has been updated accordingly.

marcopeereboom commented:
I am ok with this. Before merging this must be squashed though.

cmanv commented:
@marcopeereboom   done

LordReg commented:
@cmanv I like how the layout split works now but I found a backwards compatibility issue.

I noticed the output of +A is parsed with bar_format.  That should not be the case without a new toggle (e.g. bar_action_parse_output) with the default set to disabled.  Same goes for the output of the other character sequences such as +W.

Also, I have been considering the text markup implementation.  It's fine if the initial implementation is limited to 10 colors/fonts, but the syntax itself should be open for more.

For example, currently `+@fg24` would color '4' with bar_color index 2.
If color support is extended to remove the limit, `+@fg24` would be ambiguous.

What I am thinking is to require a terminator in the syntax and use a delimiter to clearly separate values from options.  For example,`+@fg=24;` and `+@stp;`

I suggest turning the text markup feature into a separate patch so that we can merge the new bar section feature first.

cmanv commented:
@LordReg the problem is that the parsing occurs after all character sequences have been expanded. Changing it to allow selective parsing would require a bit of work. +A is also expanded and split into sections at an early stage.

So my suggestion to keep the output of +A 'as is' by default is to set 2 toggles,

- bar_action_multi_section (default disabled): allows the definitions of multiple sections within the script's output.
- bar_text_markup_enabled (default disabled): enables text markup in the status bar.

So if a user explicitly enables text markup but doesn't want any markup within his script's output, he can add **+@stp;** before +A or at the start of the script's output. The same thing goes with +W.

Those two toggles could be implemented in short order.

Also adding a ';' to the markup sequence doesn't pose any particular problem, as long as we keep the 10 colors/fonts limit for now.

Just to make sure, do you really want a '=' before the color or font number?
**+@fg=12;**  or **+@fg12;**  ?

LordReg commented:
> @LordReg the problem is that the parsing occurs after all character sequences have been expanded. Changing it to allow selective parsing would require a bit of work. +A is also expanded and split into sections at an early stage.

A simple solution would be to escape all '+' in the expansions depending on toggle(s).  bar_replace and bar_print_layout would then skip those sequences... which leads me to a bug I just spotted in master.  '++' is not replaced with '+'.. it is copied as-is.  I will push a fix shortly.

> * bar_action_multi_section (default disabled): allows the definitions of multiple sections within the script's output.
> * bar_text_markup_enabled (default disabled): enables text markup in the status bar.
> 
> So if a user explicitly enables text markup but doesn't want any markup within his script's output, he can add **+@stp;** before +A or at the start of the script's output. The same thing goes with +W.

+| and +@ should really be usable by default since they don't break default behavior.  I am thinking it makes sense to have one new toggle:

- bar_action_expand - process bar_format character sequences in bar_action output.

It doesn't make sense to process character sequences in any other expansion besides +A.  Also, if +@ is handled in +A, then the other '+' can be handled as well.  It would be left to the user to escape any '+' that shouldn't be handled.

> Just to make sure, do you really want a '=' before the color or font number?
> **+@fg=12;** or **+@fg12;** ?

Having an '=' to separate the value from the option would provide greater flexibility.  Perhaps **+@fg=red;** could be allowed at some point, so it makes sense to me to have the value clearly separated from the option.

Also, I see the new +@ sequence as more than just text markup.  It is a general way to set options at various points in bar output and could potentially be used to set more than font/color.

cmanv commented:
New update to address LordReg's comment.

- Escape all '+' in the expansion of the window class, instance, title and workspace name.
- Escape all '+' in the output of the **bar_action** script by default as to not break default behavior.
- Add the toggle **bar_action_expand** to allow processing of **bar_format** character sequences in the action's script output. Defaults to 0.
- Use the format **+@fg=x;**, **+@bg=x;**, **+@fn=x;**. **+@stp;** for the markup sequences.
- Update the man page accordingly.

cmanv commented:
Minor fix. Remove leftover enum for bar_layout.

LordReg commented:
New updates look good.

As I understand, the *bar_font_icons* option is intended as a way for a font to be assigned only to the Unicode private use area (PUA).  According to the [Unicode site](https://www.unicode.org/charts/PDF/UE000.pdf), the range should be U+E000-F8FF instead of E000-FFFF.
Also, since the option can be used for more than an icon font, I am thinking that *bar_font_pua* would be a more fitting name.  Using it for icon fonts can be mentioned in the description.

A couple [style (9)](https://man.openbsd.org/style.9) issues that stood out:
- Second-level indents (when you split a statement to multiple lines) should be tabs followed by four spaces.
- is_valid_markup() return type should be on a line by itself preceding the function.

Once the above issues are resolved, it should be time to merge.

cmanv commented:
This update is to address @LordReg's last comment.

- Rename **bar_font_icons** to **bar_font_pua** and restrict usage to PUA code points: U+E000 -> U+F8FF.
- Fix style issues.
- Update man page.

marcopeereboom commented:
We can't merge this until conflicts are resolved.

cmanv commented:
Conflict resolved. pledge() is a no-op on my platform at this time so I cannot test it.

LordReg commented:
@cmanv Please rebase to master rather than merging.

cmanv commented:
@LordReg ok, re-done with a rebase.

marcopeereboom commented:
Doesn't even break backwards compat!!
